### PR TITLE
Test fixtures improovements

### PR DESCRIPTION
**Pull request by alpar-t**
While working on other fixtures, there's a need to do some changes to the plugin
- Support for UDP ports 
- create and correctly wore `preProcessFixture` and `postProcessFixture`  and hide `composeUp` by creating a `buildFixture` task.  These are supported on the fixture project . Post has access to the port mappings. The use-case is to  allow for setting things up for the fixture ( like config files that need to go in ) from Gradle and generate files or other resources that need to have information about ports.

### CODE REVIEW

File: buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java
Comment (nik9000) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java]: configureServiceInforForTask -> configureServiceInfoForTask

File: buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java
Comment (nik9000) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java]: Info seems a bit noisy for this.

**Comment by nik9000** [comment]
So it goes:
* preProcessFixture
* composeUp
* postProcessFixture
* buildFixture
* the test stuff
* composeDown

I can see how `postProcessFixture` is important to be able to poke things into the fixture after it is started. What is `preProcessFixture` for? To munge files that we're going to share with the container?

File: buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java
Comment (rjernst) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java]: Why even create these if it's not supported?

File: buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java
Comment (rjernst) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java]: I think we should have assurances this will occur, even if gradle gets a SIGINT. As a followup, we should spawn a child process that will catch sigint and shutdown this container (see similar work in the release manager).

File: buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java
Comment (rjernst) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java]: typo: Infor -> Info

File: buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java]: Info seems right to me. The default logging level is lifecycle, so you will not see this by default, but only when using `--info` (which is what we use in CI).

File: x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/AbstractActiveDirectoryTestCase.java
Comment (rjernst) [inline x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/AbstractActiveDirectoryTestCase.java]: nit: why is this key so long? The last "fixture" seems redundant?

**Review by rjernst — COMMENTED** [review]
I have a few minor suggestions

File: buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java]: The tasks need to exist because the build script might customize them ( like add a doLast ) . Since that runs at configuration time it would fail when docker-compose is not supported if the tasks would not be created.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java]: I think this is a bit different, since `./gradlew clean` cleans it up as `clean` depends on `composeDown`.  I was even contemplating not having this at all for this reason. 
The plugin and docker-compose are good at detecting and using existing containers and it's faster like that when you have to iterate many times on it. This is another way how this is different: new builds will not keep on adding containers.
Of course this would mean that the tests have to be written in a way that can handle that. 
I added the finalizedBy to make it resemble how it used to be but we'll need to consider how this will work with `--parallel` at some time. I propose we consider this then.

File: x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/AbstractActiveDirectoryTestCase.java
Comment (atorok) [reply x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/AbstractActiveDirectoryTestCase.java]: the second "fixture" is the name of the service from the `docker-compose.yml`. We could rename that to something else, it's needed because there can be multiple services for one docker-compose file.

**Comment by atorok** [comment]
> I can see how `postProcessFixture` is important to be able to poke things into the fixture after it is started. What is `preProcessFixture` for? To munge files that we're going to share with the container?

That's right. There's some fixtures that have to create directories up-front because the fixtures would create them with messy permissions.

**Comment by atorok** [comment]
@nik9000 @rjernst  ready for another review.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java
Comment (nik9000) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java]: So you think it is ok to leave the containers up? They certainly look like they are way lower overhead. So "best effort" seems not nearly so bad as it would be with containers.

Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java]: I don't think we should leave containers running. Whether or not they are rebuilt/reused is a different matter. But when the build finishes, regardless of whether it finished normally or was Ctrl-C'd, there should be nothing running (granted this is not exactly true because of the gradle daemon, but at least that is something that will destroy itself on a timer).

File: x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/AbstractActiveDirectoryTestCase.java
Comment (rjernst) [reply x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/AbstractActiveDirectoryTestCase.java]: Why not have "smb-fixture" be the name of the service in docker-compose? ie:
`test.fixtures.smb-fixture.tcp`

**Comment by rjernst** [comment]
@atorok I left some more replies on the original comments.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixturesPlugin.java]: OK, I'we put it on my TODO list to create a wrapper script to make sure containers are downed even if Gradle doesn't live long enough to run the finalizers.

**Comment by atorok** [comment]
thanks @rjernst I implemented your suggestion for the property notes and will do the cleanup of containers in a follow up.

**Comment by rjernst** [comment]
I do not think this should be checked in without protections for container cleanup. Leaving anything running is unexpected and broken behavior. The build should not leave extraneous running processes which a developer must manually cleanup.

**Comment by atorok** [comment]
@rjernst starting the containers is already checked in. I don't think there's anything in this PR which makes that any worse, so I don't think we should hold up this one.

**Comment by rjernst** [comment]
I had thought this added new uses, but I see I was mistaken. I'm fine pushing this then, but I think the work to protect against runaway external processes needs to be high priority.

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by atorok** [comment]
Thanks, ACK on the high priority.

**Pull request merged**
